### DISCUSSION TITLE
Logging

### DISCUSSION THREAD

**Discussion by votkon** [ajndkr/lanarky#65] [General]
That would be great to have chat logging implemented somehow.
Currently, it is impossible(?) to save the **chain.memory.chat_memory.messages** on websocket disconnection.
Websocket disconnect doesn't seem to raise any exceptions when closed correctly.

**Comment by ajndkr** [comment]
Hey! can you please elaborate what you mean by logging here. Do you mean logging messages into a database or simply logging on the terminal when you run your app?

**Reply by votkon** [reply]
Hey, 
I meant just catching the user disconnect event to push the conversation somewhere(DB, logs etc).
It looks like I found a solution for this:

```
@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket, chain: ConversationChain = Depends(conversation_chain)):
    connection = WebsocketConnection.from_chain(chain=chain, websocket=websocket)
    try:
        await connection.connect()
    except Exception as e:
        handle_closed_connection(chain.memory.chat_memory)
        return
    else:
        handle_closed_connection(chain.memory.chat_memory)

```
This passes the chain's chat history to the _handle_closed_connection_ function for further treatment.
Thanks!

**Reply by ajndkr** [reply]
Gotcha! Looks like a good idea. Needs to be a bit more generic I think depending on individual use case. Maybe a `BackgroundTask` can be used here. I will investigate further.
